src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs

```diff
@@ -123,6 +123,22 @@ public void SetOptions(string commandLineForOptions)
             }
         }
 
+        public string DefaultNamespace
+        {
+            get => _visualStudioProject.DefaultNamespace;
+            private set => _visualStudioProject.DefaultNamespace = value;
+        }
+
+        public void SetProperty(string name, object value)
+        {
+            if (string.Equals(name, AdditionalPropertyNames.DefaultNamespace, StringComparison.OrdinalIgnoreCase) 
+                && _visualStudioProject.Language == LanguageNames.CSharp
```

Comment:
Done

Reply:
Except we also want this for VB too...please figure out how that design works here as well. :smile:

Reply:
Unless we want to add a `RootNamespace` property to `Project` and set that to the exact same value user can get from VB compilation options or `null` for non-VB projects, I'd stick with current logic for now.

I'm not saying the proposal above is bad, we definitely should discuss how default/root namespace should work for all roslyn workspace projects at some point. But it's not a crucial part for this change.

Reply:
@genlu It's not strictly necessary, but my concern is if we're designing this into a corner here without thinking bigger picture. That, and these types are really language-agnostic and so any language checking *at all* is a very big code smell.

Reply:
I get what you are saying. But my gut feeling is we would end up not support "default namespace" to VB, I don't really see a way to design it so that root and default namespace could co-exist in a project without being super confusing. Plus VB users seem to be less likely to declare namespace in source (https://github.com/dotnet/vblang/issues/351), so I'm not sure there's much demand for this. Thoughts?

Reply:
my main issue here as well is that this is hardocding C# behaviro into the common layer.  if we need c# specific behavior, it should go at teh c# level.  Or, if we want common behavior for everything but VB, then the exception shuld go in the VB layer.

For all intensive porpoises, the common layer should not reference LanguageNames.CSharp or LanguageNames.VB.  (only exception is when the common layer is truly making a component common to both and it needs to reference both names to export it.

Reply:
I agree with you. But we don't seem to have a language specific layer for CPS project (like what we have for legacy projects). @jasonmalinowski Any suggestions?

Reply:
Well, if we just set the value to whatever project system provides us, then it would make no sense for VB (not sure about F#) as the default namespace property of the project will then have a non-null value. 

One way I think we can do is, instead of having a concept of default namespace in workspace that is C# only, we change it to `RootNamespace` which just represents the value of `rootnamespace` property in project file. Then we will let each individual feature do the language check and decide how to interpreter this value.

Reply:
Based on offline discussion with @jasonmalinowski, now we just expose root namespace directly, and will let the individual feature decide what to do with it.